app/routes/api.gitlab.commits.ts

```diff
@@ -0,0 +1,139 @@
+import { type ActionFunctionArgs, json } from '@remix-run/cloudflare';
+import { GitlabService } from '~/lib/persistenceGitbase/gitlabService';
+import type { GitlabProject } from '~/lib/persistenceGitbase/types';
+import { withV8AuthUser } from '~/lib/verse8/middleware';
+
+export const action = withV8AuthUser(commitsAction, { checkCredit: true });
+export const loader = withV8AuthUser(commitsLoader, { checkCredit: true });
+
+/**
+ * Loader function for getting project commits
+ */
+async function commitsLoader({ context, request }: ActionFunctionArgs) {
+  const env = { ...context.cloudflare.env, ...process.env } as Env;
+
+  const url = new URL(request.url);
+  const projectPath = url.searchParams.get('projectPath');
+  const page = url.searchParams.get('page') || '1';
+  const perPage = url.searchParams.get('perPage') || '20';
+  const branch = url.searchParams.get('branch') || '';
+
+  if (!projectPath) {
+    return json({ success: false, message: 'Project path is required' }, { status: 400 });
+  }
+
+  const parsedPage = parseInt(page, 10);
+  const parsedPerPage = parseInt(perPage, 10);
+
+  // Use undefined for empty branch parameter
+  const branchToUse = branch === '' ? undefined : branch;
+
+  const gitlabService = new GitlabService(env);
+
+  try {
+    const result = await gitlabService.getProjectCommits(projectPath, parsedPage, parsedPerPage, branchToUse);
+
+    return json({
+      success: true,
+      data: {
+        project: result.project,
+        commits: result.commits,
+        pagination: {
+          total: result.total,
+          page: parsedPage,
+          perPage: parsedPerPage,
+          hasMore: result.hasMore,
+          totalPages: Math.ceil(result.total / parsedPerPage),
+        },
+      },
+    });
+  } catch (error) {
+    const errorMessage = error instanceof Error ? error.message : 'Unknown error';
+    return json({ success: false, message: `Failed to fetch project commits: ${errorMessage}` }, { status: 500 });
+  }
+}
+
+/**
+ * Action function for committing files to project
+ */
+async function commitsAction({ context, request }: ActionFunctionArgs) {
+  const env = { ...context.cloudflare.env, ...process.env } as Env;
+  const user = context?.user as { email: string; isActivated: boolean };
+
+  const gitlabService = new GitlabService(env);
+
+  if (!user) {
+    return json({ success: false, message: 'Authentication required' }, { status: 401 });
+  }
+
+  if (!user.isActivated) {
+    return json({ success: false, message: 'User is not activated' }, { status: 403 });
+  }
+
+  const email = user.email;
+
+  // JSON 데이터로 받기
+  const requestData = (await request.json()) as {
+    projectName: string;
+    isFirstCommit: boolean;
+    description?: string;
+    commitMessage: string;
+    branch?: string;
+    files: {
+      path: string;
+      content: string;
+    }[];
+  };
```

Comment:
How about turn this (and similar things) to `interface`?

Reply:
Agreed, but it’s not duplicated, tolerable for now